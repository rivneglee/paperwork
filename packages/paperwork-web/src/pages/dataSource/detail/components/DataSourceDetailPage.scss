.pwapp-field-grants-table {
  @media (min-width: 768px) {
    width: 400px;
  }
}
