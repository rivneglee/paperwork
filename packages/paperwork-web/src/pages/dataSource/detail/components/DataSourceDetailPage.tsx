import React, { FunctionComponent, useState } from 'react';
import {
  BaseTemplate,
  Card,
  Drawer,
  Input, Scrollable,
  Separator,
} from '@paperwork/ui-widgets';

import AppBar from '../../../../components/AppBar/AppBar';
import { DataSource, Field, Grant } from '../../../../schema/DataSource';
import DataSourceDetailModal from './DataSourceDetailModal';
import DataSourceDetailActions from './DataSourceDetailActions';
import FieldsTable from './FieldsTable';
import FieldGrantsTable from './FieldGrantsTable';

import './DataSourceDetailPage.scss';

interface Props {
  dataSource: DataSource;
  isPageEdited: boolean;
  onUpdateDetail: (key: string, value: any) => void;
  onUpdateField: (index: number, key: string, value: string) => void;
  onAddField: (newField: Field, key: string, value: any) => void;
  onRemoveField: (index: number) => void;
  onSave: (dataSource: DataSource) => void;
  onDelete: () => void;
  onCancel: () => void;
  onEditGrant: (fieldId?: string) => void;
  grantField: Field;
  onAddGrant: (newGrant: Grant, key: string, value: any) => void;
  onUpdateGrant: (index: number, key: string, value: string) => void;
  onRemoveGrant: (index: number) => void;
}

const DataSourceDetailPage: FunctionComponent<Props> = ({
  dataSource,
  isPageEdited,
  onUpdateDetail,
  onAddField,
  onUpdateField,
  onRemoveField,
  onSave,
  onCancel,
  onDelete,
  onEditGrant,
  onAddGrant,
  onUpdateGrant,
  onRemoveGrant,
  grantField,
}) => {
  const [modalType, setModalType] = useState('');

  const { id } = dataSource;

  const isCreating = !id;

  const onCloseModal = () => setModalType('');

  const onClickCancel = () => {
    if (isPageEdited) {
      setModalType('unsave');
    } else {
      onCancel();
    }
  };

  const onClickSave = () => onSave(dataSource);

  const onClickDelete = () => setModalType('delete');

  const onCloseDrawer = () => onEditGrant();

  return (
    <BaseTemplate
      header={<AppBar />}
    >
      <Card
        header={
          <h3>Datasource details</h3>
        }>
        <Input
          value={dataSource.name}
          label="Datasource name"
          isRequired
          labelPlacement="top"
          size="s"
          onChange={(e: any) => onUpdateDetail('name', e.target.value)}
        />
        <Separator/>
        <FieldsTable
          onAddField={onAddField}
          onUpdateField={onUpdateField}
          onRemoveField={onRemoveField}
          onEditGrant={onEditGrant}
          fields={dataSource.fields}
        />
      </Card>
      <DataSourceDetailActions
        isCreating={isCreating}
        onClickCancel={onClickCancel}
        onClickDelete={onClickDelete}
        onClickSave={onClickSave}
      />
      { modalType && (
          <DataSourceDetailModal
            modalType={modalType}
            onCloseModal={onCloseModal}
            onCancel={onCancel}
            onDelete={onDelete}
          />
       )}
      <Drawer
        placement="right"
        header={<h3>Permission</h3>}
        isShow={!!grantField}
        onClose={onCloseDrawer}
      >
        <Scrollable className="pwapp-field-grants-table">
          {grantField && (
            <FieldGrantsTable
              field={grantField}
              onAddGrant={onAddGrant}
              onUpdateGrant={onUpdateGrant}
              onRemoveGrant={onRemoveGrant}
            />
          )}
        </Scrollable>
      </Drawer>
    </BaseTemplate>
  );
};

export default DataSourceDetailPage;
