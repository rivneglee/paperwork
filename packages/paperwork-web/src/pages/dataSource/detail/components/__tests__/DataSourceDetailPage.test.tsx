import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import DataSourceDetailPage from '../DataSourceDetailPage';

describe('DataSourceDetailPage', () => {
  const dataSource = {
    id: '1',
    name: 'Engagement survey 2020',
    owner: {
      id: '1',
      displayName: 'Alex Li',
      username: 'alexli',
    },
    fields: [
      {
        id: '123',
        name: 'STAFF_NAME',
        grants: [],
      },
    ],
  };

  const commonProps = {
    isPageEdited: true,
    onUpdateDetail: jest.fn(),
    onUpdateField: jest.fn(),
    onAddField: jest.fn(),
    onRemoveField: jest.fn(),
    onSave: jest.fn(),
    onDelete: jest.fn(),
    onCancel: jest.fn(),
    onEditGrant: jest.fn(),
    onAddGrant: jest.fn(),
    onUpdateGrant: jest.fn(),
    onRemoveGrant: jest.fn(),
  };

  it('should render properly if logged in as owner', () => {
    const wrapper = shallow(
      <DataSourceDetailPage dataSource={dataSource} {...commonProps} isOwner={true} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render properly if not datasource owner', () => {
    const wrapper = shallow(
      <DataSourceDetailPage dataSource={dataSource} {...commonProps} isOwner={false} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render properly when grantField is given', () => {
    const wrapper = shallow(
      <DataSourceDetailPage
        grantField={{
          id: '123',
          name: 'STAFF_NAME',
          grants: [],
        }}
        dataSource={dataSource}
        {...commonProps}
        isOwner={true}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
