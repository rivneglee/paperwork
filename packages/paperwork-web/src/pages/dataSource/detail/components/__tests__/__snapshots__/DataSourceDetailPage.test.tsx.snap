// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSourceDetailPage should render properly if logged in as owner 1`] = `
<Di
  header={
    <Connect(AppBar)
      activeMenuId="datasource"
    />
  }
  spinner={<Unknown />}
>
  <i
    header={
      <Unknown
        primary={
          <Unknown>
            Datasource details 
          </Unknown>
        }
      />
    }
  >
    <t
      disabled={false}
      isRequired={true}
      label="Datasource name"
      labelPlacement="top"
      onChange={[Function]}
      size="s"
      value="Engagement survey 2020"
    />
    <x />
    <OwnerFieldsTable
      fields={
        Array [
          Object {
            "grants": Array [],
            "id": "123",
            "name": "STAFF_NAME",
          },
        ]
      }
      onAddField={[Function]}
      onEditGrant={[Function]}
      onRemoveField={[Function]}
      onUpdateField={[Function]}
    />
  </i>
  <DataSourceDetailActions
    isCreating={false}
    isOwner={true}
    onClickCancel={[Function]}
    onClickDelete={[Function]}
    onClickSave={[Function]}
  />
  <m
    header={
      <h3>
        Permission
      </h3>
    }
    isShow={false}
    onClose={[Function]}
    placement="right"
  >
    <b
      className="pwapp-field-grants-table"
    />
  </m>
</Di>
`;

exports[`DataSourceDetailPage should render properly if not datasource owner 1`] = `
<Di
  header={
    <Connect(AppBar)
      activeMenuId="datasource"
    />
  }
  spinner={<Unknown />}
>
  <i
    header={
      <Unknown
        primary={
          <Unknown>
            Datasource details 
            <z
              color="secondary"
            >
              Collaborative
            </z>
          </Unknown>
        }
      />
    }
  >
    <t
      disabled={true}
      isRequired={true}
      label="Datasource name"
      labelPlacement="top"
      onChange={[Function]}
      size="s"
      value="Engagement survey 2020"
    />
    <x />
    <CollaboratorFieldsTable
      fields={
        Array [
          Object {
            "grants": Array [],
            "id": "123",
            "name": "STAFF_NAME",
          },
        ]
      }
    />
  </i>
  <DataSourceDetailActions
    isCreating={false}
    isOwner={false}
    onClickCancel={[Function]}
    onClickDelete={[Function]}
    onClickSave={[Function]}
  />
  <m
    header={
      <h3>
        Permission
      </h3>
    }
    isShow={false}
    onClose={[Function]}
    placement="right"
  >
    <b
      className="pwapp-field-grants-table"
    />
  </m>
</Di>
`;

exports[`DataSourceDetailPage should render properly when grantField is given 1`] = `
<Di
  header={
    <Connect(AppBar)
      activeMenuId="datasource"
    />
  }
  spinner={<Unknown />}
>
  <i
    header={
      <Unknown
        primary={
          <Unknown>
            Datasource details 
          </Unknown>
        }
      />
    }
  >
    <t
      disabled={false}
      isRequired={true}
      label="Datasource name"
      labelPlacement="top"
      onChange={[Function]}
      size="s"
      value="Engagement survey 2020"
    />
    <x />
    <OwnerFieldsTable
      fields={
        Array [
          Object {
            "grants": Array [],
            "id": "123",
            "name": "STAFF_NAME",
          },
        ]
      }
      onAddField={[Function]}
      onEditGrant={[Function]}
      onRemoveField={[Function]}
      onUpdateField={[Function]}
    />
  </i>
  <DataSourceDetailActions
    isCreating={false}
    isOwner={true}
    onClickCancel={[Function]}
    onClickDelete={[Function]}
    onClickSave={[Function]}
  />
  <m
    header={
      <h3>
        Permission
      </h3>
    }
    isShow={true}
    onClose={[Function]}
    placement="right"
  >
    <b
      className="pwapp-field-grants-table"
    >
      <FieldGrantsTable
        field={
          Object {
            "grants": Array [],
            "id": "123",
            "name": "STAFF_NAME",
          }
        }
        onAddGrant={[Function]}
        onRemoveGrant={[Function]}
        onUpdateGrant={[Function]}
      />
    </b>
  </m>
</Di>
`;
